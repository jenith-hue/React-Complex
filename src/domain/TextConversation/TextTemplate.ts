export interface QuickResponse {
  title: string;
  message: string;
  spanishMessage: string;
}

export interface TextTemplate {
  category: string;
  order: number;
  quickresponse: QuickResponse[];
}
