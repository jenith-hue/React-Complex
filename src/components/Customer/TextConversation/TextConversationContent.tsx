import {
  makeStyles,
  RACCOLOR,
  RACSelect,
  RACTextConversationMessage,
  RACTextMessageInput,
  Typography,
} from '@rentacenter/racstrap';
import React, { useEffect, useState, useRef } from 'react';
import clsx from 'clsx';
import {
  useTextConversation,
  useTextConversationActions,
} from '../../../context/TextConversationProvider/TextConversationProvider';
import { TextTemplateModal } from '../TextTemplate/TextTemplateModal';
import { ApiStateWrapper } from '../../../common/ApiStateWrapper/ApiStateWrapper';
import { formatPhoneNumber } from '../../../utils/utils';
import { useStoreDetails } from '../../../context/Store/StoreProvider';
import { useUserPermissions } from '../../../context/permission/PermissionsProvider';
import { useUserStateContext } from '../../../context/user/user-contexts';
import { FRCStoreType, RentACenter } from '../../../constants/constants';

export interface TextConversationContentProps {
  onComponentDidMount: () => void;
}

const useStyles = makeStyles((theme: any) => ({
  root: {
    padding: '1rem 0.5rem 1rem 0.5rem',
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
  },
  header: {
    display: 'flex',
    width: '100%',
    // eslint-disable-next-line sonarjs/no-duplicate-string
    justifyContent: 'flex-start',
  },
  body: {
    width: '100%',
    overflowY: 'auto',
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-end',
  },
  noRecord: {
    display: 'flex',
    justifyContent: 'center !important',
  },
  headerItem: {
    display: 'flex',
    alignItems: 'flex-start',
    flexDirection: 'column',
    flex: 1,
  },
  headerItemValue: {
    color: `${RACCOLOR.INDEPENDENCE} !important`,
    whiteSpace: 'nowrap',
  },
  centered: {
    alignItems: 'center',
  },
  messageDate: {
    fontSize: `${theme.typography.pxToRem(12)} !important`,
    color: `${RACCOLOR.ARGENT} !important`,
    marginTop: '1rem',
    marginBottom: '1rem',
  },
  leftMargin: {
    marginLeft: '2rem',
  },
}));
export const TextConversationContent = ({
  onComponentDidMount,
}: TextConversationContentProps) => {
  const classes = useStyles();
  const ref = useRef(null);
  const {
    phoneOptions,
    selectedPhoneOption,
    fullName,
    messages,
    isLoading,
    hasApiError,
    preferredLanguage,
    isCommunicationAllowed,
  } = useTextConversation();

  const { onPhoneOptionChanged, sendMessage } = useTextConversationActions();
  const [openTextTemplateModal, setOpenTextTemplateModal] = useState(false);
  const { store } = useStoreDetails();
  const { user } = useUserStateContext();
  const CoworkerName = `${user?.firstName} ${user?.lastName
    .charAt(0)
    .toUpperCase()}`;
  const companyName =
    store?.storeType === FRCStoreType
      ? `${CoworkerName} ${store?.storeNumber}`
      : RentACenter;
  const { hasAccessToTextTemplate, hasAccessToFreeFormText } =
    useUserPermissions();
  const DEFAULT_PRE_INPUT_TEXT = 'Templates';
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(onComponentDidMount, []);

  useEffect(() => {
    if (ref?.current) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      //@ts-ignore
      ref?.current?.scroll({
        top: Number.MAX_SAFE_INTEGER,
        left: 0,
        behavior: 'smooth',
      });
    }
  }, [messages, isLoading]);

  return (
    <div className={classes.root}>
      <div
        className={clsx(
          classes.body,
          isLoading || !messages.length || hasApiError ? classes.noRecord : null
        )}
        ref={ref}
      >
        <ApiStateWrapper
          loading={isLoading}
          hasApiError={hasApiError}
          response={messages}
          successContent={
            <>
              {messages.map(
                ({
                  textMessageId,
                  isSentByRAC,
                  message,
                  messageDate,
                  initials,
                  tooltipTitle,
                  tooltipSubTitle,
                }) => (
                  <div key={textMessageId}>
                    <Typography
                      align="center"
                      variant="body1"
                      className={classes.messageDate}
                    >
                      {messageDate}
                    </Typography>
                    <RACTextConversationMessage
                      key={textMessageId}
                      isOwnMessage={isSentByRAC}
                      avatarText={initials}
                      message={message}
                      tooltipTitle={tooltipTitle}
                      tooltipSubtitle={tooltipSubTitle}
                    />
                  </div>
                )
              )}
            </>
          }
        />
      </div>
      <div className={classes.header}>
        <div className={classes.headerItem}>
          <Typography align="left" variant="h5">
            Name
          </Typography>
          <Typography
            className={classes.headerItemValue}
            variant="body1"
            align="left"
          >
            {fullName}
          </Typography>
        </div>
        <div className={clsx(classes.headerItem, classes.leftMargin)}>
          <Typography align="left" variant="h5">
            Phone
          </Typography>
          <RACSelect
            isDisabled={!isCommunicationAllowed || !phoneOptions?.length}
            defaultValue={selectedPhoneOption?.value}
            options={phoneOptions.map((phone) => ({
              ...phone,
              label: formatPhoneNumber(phone.label, !isCommunicationAllowed),
            }))}
            onChange={(e: React.ChangeEvent<{ value: any }>) => {
              onPhoneOptionChanged(e.target.value);
            }}
          />
        </div>

        <div className={clsx(classes.headerItem, classes.centered)}>
          <Typography align="center" variant="h5">
            Type
          </Typography>
          <Typography
            align="center"
            className={classes.headerItemValue}
            variant="body1"
          >
            {selectedPhoneOption?.phoneType}
          </Typography>
        </div>
        <div className={classes.headerItem}>
          <Typography align="center" variant="h5">
            Language
          </Typography>
          <Typography
            align="center"
            className={classes.headerItemValue}
            variant="body1"
          >
            {preferredLanguage}
          </Typography>
        </div>
      </div>
      {hasAccessToTextTemplate && hasAccessToFreeFormText && (
        <div>
          <Typography variant="h5">Send Text To:</Typography>
          <Typography variant="body1">
            Choose Text Template or Free Form Text
          </Typography>
        </div>
      )}
      <TextTemplateModal
        open={openTextTemplateModal}
        onClose={setOpenTextTemplateModal}
        preferredLanguageDesc={preferredLanguage}
        isCommunicationAllowed={isCommunicationAllowed}
      />
      <RACTextMessageInput
        onSend={(e) => {
          sendMessage(`${e} - ${companyName}`);
        }}
        onPreInputTextClicked={() =>
          hasAccessToTextTemplate && setOpenTextTemplateModal(true)
        }
        isCommunicationAllowed={
          isCommunicationAllowed &&
          hasAccessToFreeFormText &&
          phoneOptions?.length > 0
        }
        preInputText={
          hasAccessToTextTemplate && phoneOptions?.length > 0
            ? DEFAULT_PRE_INPUT_TEXT
            : ' '
        }
      />
    </div>
  );
};
