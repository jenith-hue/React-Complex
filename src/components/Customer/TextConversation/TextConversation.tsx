import { faSync } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  Icon,
  makeStyles,
  RACCard,
  RACTabs,
  Typography,
} from '@rentacenter/racstrap';
import React from 'react';
import { CO_CUSTOMER_TAB, CUSTOMER_TAB } from '../../../constants/constants';
import {
  AVAILABLE_TABS,
  useTextConversation,
  useTextConversationActions,
} from '../../../context/TextConversationProvider/TextConversationProvider';
import { TextConversationContent } from './TextConversationContent';

export const textConversationDataTestid = 'textConversationDataTestid';

const useStyles = makeStyles((theme: any) => ({
  textConversationRoot: {
    marginTop: '2rem',
    marginBottom: '1rem',
  },
  row: {
    display: 'flex',
    width: '100%',
    justifyContent: 'space-between',
  },
  cardContentContainer: {
    marginLeft: '1.25rem',
    marginRight: '1.25rem',
  },
  tabPanel: {
    height: '26rem',
    maxHeight: '26rem',
    overflow: 'auto',
  },
  icon: {
    color: theme.palette.primary.main,
    cursor: 'pointer',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    fontSize: theme.typography.pxToRem(20),
    position: 'absolute',
    right: 0,
    height: '100%',
  },
}));
export const TextConversation = () => {
  const classes = useStyles();

  const { disabledTabs, customerPhones, coCustomerPhones } =
    useTextConversation();
  const { onRefresh, setTextConversationSelectedTab } =
    useTextConversationActions();

  if (!customerPhones?.length && !coCustomerPhones?.length) return null;

  return (
    <div
      className={classes.textConversationRoot}
      data-testid={textConversationDataTestid}
    >
      <div className={classes.row}>
        <Typography variant="h4">Text Conversation</Typography>
      </div>
      <RACCard>
        <div className={classes.cardContentContainer}>
          <RACTabs
            classes={{
              tabPanel: classes.tabPanel,
            }}
            icon={
              <Icon onClick={onRefresh} classes={{ root: classes.icon }}>
                <FontAwesomeIcon icon={faSync} />
              </Icon>
            }
            defaultValue={0}
            tabs={AVAILABLE_TABS}
            contentForTabs={[
              <TextConversationContent
                key="customer"
                onComponentDidMount={() => {
                  setTextConversationSelectedTab(CUSTOMER_TAB);
                }}
              />,
              <TextConversationContent
                key="co-customer"
                onComponentDidMount={() =>
                  setTextConversationSelectedTab(CO_CUSTOMER_TAB)
                }
              />,
            ]}
            disabledTabs={disabledTabs}
          />
        </div>
      </RACCard>
    </div>
  );
};
