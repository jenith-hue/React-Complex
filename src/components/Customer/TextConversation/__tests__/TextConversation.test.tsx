/* eslint-disable react/display-name */
/* eslint-disable sonarjs/no-identical-functions */

import React from 'react';
import renderer from 'react-test-renderer';
import { TextConversation } from '../TextConversation';

interface PropWithChildren {
  children: JSX.Element;
}

jest.mock('@rentacenter/racstrap', () => ({
  Typography: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Typography" {...rest}>
      {children}
    </div>
  ),
  RACCard: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACCard" {...rest}>
      {children}
    </div>
  ),
  RACTabs: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTabs" {...rest}>
      {children}
    </div>
  ),
  Icon: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Icon" {...rest}>
      {children}
    </div>
  ),
  RACCOLOR: {},
  makeStyles: () => () => ({
    textConversationRoot: 'textConversationRoot',
    row: 'row',
    cardContentContainer: 'cardContentContainer',
    appBar: 'appBar',
    tabPanel: 'tabPanel',
    icon: 'icon',
  }),
}));

jest.mock('../TextConversationContent', () => ({
  TextConversationContent: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="TextConversationContent" {...rest}>
      {children}
    </div>
  ),
}));

jest.mock(
  '../../../../context/TextConversationProvider/TextConversationProvider',
  () => ({
    useTextConversationActions: () => ({
      onCustomerTabSelected: jest.fn(),
      onCoCustomerTabSelected: jest.fn(),
      onRefresh: jest.fn(),
    }),
    useTextConversation: () => ({
      disabledTabs: [],
      customerPhones: [
        {
          id: 'id1',
        },
      ],
      coCustomerPhones: [
        {
          id: 'id1',
        },
      ],
      isCommunicationAllowed: true,
      isCommunicationAllowedWithCoCustomer: true,
    }),
  })
);

jest.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="FontAwesomeIcon" {...rest}>
      {children}
    </div>
  ),
}));
jest.mock(
  '../../../../context/CustomerDetails/CustomerDetailsProvider',
  () => ({
    useCustomerDetailsActions: () => ({
      fetchCommunicationDetailsForCustomer: jest.fn(),
    }),
    useCustomerDetails: () => ({
      customerDetails: {
        coCustomerPhones: [
          {
            id: 'id1',
          },
        ],
        personalReferences: [
          {
            id: 'id1',
          },
        ],
        employerReferences: [
          {
            id: 'id1',
          },
        ],
      },
      coCustomerDetails: {
        phones: [
          {
            id: 'id1',
          },
        ],
      },
    }),
  })
);
describe('TextConversation', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<TextConversation />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
