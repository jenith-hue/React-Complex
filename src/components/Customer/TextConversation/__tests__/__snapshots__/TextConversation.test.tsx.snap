// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextConversation renders correctly 1`] = `
<div
  className="textConversationRoot"
  data-testid="textConversationDataTestid"
>
  <div
    className="row"
  >
    <div
      data-testid="Typography"
      variant="h4"
    >
      Text Conversation
    </div>
  </div>
  <div
    data-testid="RACCard"
  >
    <div
      className="cardContentContainer"
    >
      <div
        classes={
          Object {
            "tabPanel": "tabPanel",
          }
        }
        contentForTabs={
          Array [
            <TextConversationContent
              onComponentDidMount={[Function]}
            />,
            <TextConversationContent
              onComponentDidMount={[Function]}
            />,
          ]
        }
        data-testid="RACTabs"
        defaultValue={0}
        disabledTabs={Array []}
        icon={
          <Icon
            classes={
              Object {
                "root": "icon",
              }
            }
            onClick={[MockFunction]}
          >
            <FontAwesomeIcon
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f021",
                    "M440.65 12.57l4 82.77A247.16 247.16 0 0 0 255.83 8C134.73 8 33.91 94.92 12.29 209.82A12 12 0 0 0 24.09 224h49.05a12 12 0 0 0 11.67-9.26 175.91 175.91 0 0 1 317-56.94l-101.46-4.86a12 12 0 0 0-12.57 12v47.41a12 12 0 0 0 12 12H500a12 12 0 0 0 12-12V12a12 12 0 0 0-12-12h-47.37a12 12 0 0 0-11.98 12.57zM255.83 432a175.61 175.61 0 0 1-146-77.8l101.8 4.87a12 12 0 0 0 12.57-12v-47.4a12 12 0 0 0-12-12H12a12 12 0 0 0-12 12V500a12 12 0 0 0 12 12h47.35a12 12 0 0 0 12-12.6l-4.15-82.57A247.17 247.17 0 0 0 255.83 504c121.11 0 221.93-86.92 243.55-201.82a12 12 0 0 0-11.8-14.18h-49.05a12 12 0 0 0-11.67 9.26A175.86 175.86 0 0 1 255.83 432z",
                  ],
                  "iconName": "sync",
                  "prefix": "fas",
                }
              }
            />
          </Icon>
        }
      />
    </div>
  </div>
</div>
`;
